internal/controller/deployment_ to_pdb_controller.go

```diff
@@ -0,0 +1,166 @@
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// DeploymentToPDBReconciler reconciles a Deployment object and ensures an associated PDB is created and deleted
+type DeploymentToPDBReconciler struct {
+	client.Client
+	Scheme   *runtime.Scheme
+	Recorder record.EventRecorder
+}
+
+// Reconcile watches for Deployment changes (created, updated, deleted) and creates or deletes the associated PDB.
+// creates pdb with minAvailable to be same as replicas for any deployment
+func (r *DeploymentToPDBReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+
+	// Fetch the Deployment instance
+	var deployment v1.Deployment
+	if err := r.Get(ctx, req.NamespacedName, &deployment); err != nil {
+		log.Error(err, "unable to fetch Deployment")
+
+		_, e := r.handleDeploymentDeletion(ctx, req)
+		if e != nil {
+			return ctrl.Result{}, e
+		}
+		return reconcile.Result{}, client.IgnoreNotFound(err)
+	}
+	log.Info("Found: ", "deployment", deployment.Name, "namespace", deployment.Namespace)
+	// If the Deployment is created, ensure a PDB exists
+	return r.handleDeploymentCreation(ctx, &deployment)
+}
+
+// handleDeploymentCreation creates a PodDisruptionBudget when a Deployment is created or updated.
+func (r *DeploymentToPDBReconciler) handleDeploymentCreation(ctx context.Context, deployment *v1.Deployment) (reconcile.Result, error) {
+	log := log.FromContext(ctx)
+	// Check if PDB already exists for this Deployment
+	pdb := &policyv1.PodDisruptionBudget{}
+	err := r.Get(ctx, client.ObjectKey{
+		Namespace: deployment.Namespace,
+		Name:      deployment.Name + "-pdb",
+	}, pdb)
+
+	if err == nil {
+		// PDB already exists, nothing to do
+		log.Info("PodDisruptionBudget already exists", "namespace", deployment.Namespace, "name", deployment.Name)
+		//if pdb.Spec.MinAvailable.IntVal != *deployment.Spec.Replicas {
+		//	pdb.Spec.MinAvailable.IntVal = *deployment.Spec.Replicas
+		//	if err := r.Update(ctx, pdb); err != nil {
+		//		return reconcile.Result{}, err
+		//	}
+		//}
+		return reconcile.Result{}, nil
+	}
+
+	// Create a new PDB for the Deployment
+	pdb = &policyv1.PodDisruptionBudget{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "PodDisruptionBudget",
+			APIVersion: "policy/v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      r.generatePDBName(deployment.Name),
+			Namespace: deployment.Namespace,
+			Annotations: map[string]string{
+				"createdBy": "DeploymentToPDBController",
+				"target":    deployment.Name,
+			},
+		},
+		Spec: policyv1.PodDisruptionBudgetSpec{
+			MinAvailable: &intstr.IntOrString{IntVal: *deployment.Spec.Replicas},
+			Selector:     &metav1.LabelSelector{MatchLabels: deployment.Spec.Selector.MatchLabels},
```

Comment:
so there could be another pdb that selects same deployment pods and api server will actually error on any evictions for that. So we should probably list pdbs in this namespae and make sure none of them would select the deployments pods already? (have to compare each's selectors against the deployment.template.spec label. 
(can show you the code in api server if that helps we handle the error type in our cordon and drain

Reply:
if you don't want to handle that in this PR make an issue for it.